File: WebClient SDK/src/default/forums/krSettingsForumsSettingsUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { DefaultCardTypes } from 'tessa/workflow';
import { createTypedField, DotNetType } from 'tessa/platform';
import { CardRow, CardRowsListener } from 'tessa/cards';
import { GridRowAction, GridViewModel } from 'tessa/ui/cards/controls';
import { ForumHelper } from 'tessa/forums';
import { WfUiHelper } from '../workflow/wf/wfUiHelper';

export class KrSettingsForumsSettingsUIExtension extends CardUIExtension {
  private _cardRowsListener: CardRowsListener | null;
  private _disposers: Array<() => void> = [];

  public initialized(context: ICardUIExtensionContext) {
    const { model } = context;

    if (model.inSpecialMode) {
      return;
    }

    const cardTypeId = model.card.typeId;

    if (cardTypeId === DefaultCardTypes.KrSettingsTypeID) {
      const sections = model.card.tryGetSections();

      const cardTypesSection = sections?.tryGet('KrSettingsCardTypes');
      const cardTypesControl = model.controls.get('CardTypeControl');

      if (cardTypesSection && cardTypesControl) {
        const cardTypesGrid = cardTypesControl as GridViewModel;

        cardTypesGrid.rowInvoked.add(e => {
          if (e.action === GridRowAction.Inserted || e.action === GridRowAction.Opening) {
            const forumsBlock = e.rowModel?.blocks.get('UseForumBlock');

            if (forumsBlock) {
              const useForumFirst = e.row.get(ForumHelper.UseForumField) as boolean;

              WfUiHelper.setControlVisibility(
                forumsBlock,
                ForumHelper.UseForumSuffix,
                useForumFirst
              );

              const dispose = e.row.fieldChanged.addWithDispose(e => {
                if (e.fieldName === ForumHelper.UseForumField) {
                  const useForum = e.fieldValue as boolean;
                  e.storage.set(
                    ForumHelper.UseDefaultDiscussionTabField,
                    createTypedField(useForum, DotNetType.Boolean)
                  );
                  WfUiHelper.setControlVisibility(
                    forumsBlock,
                    ForumHelper.UseForumSuffix,
                    useForum
                  );
                }
              });

              forumsBlock.form.closed.addOnce(() => {
                if (dispose) {
                  dispose();
                }
              });
            }
          }
        });

        for (const cardTypeRow of cardTypesSection.rows) {
          this.attachHandlersToCardTypeRow(cardTypeRow);
        }

        this._cardRowsListener = new CardRowsListener();
        this._cardRowsListener.rowInserted.add((_s, row) => {
          this.attachHandlersToCardTypeRow(row);
        });
        this._cardRowsListener.start(cardTypesSection.rows);
      }
    } else if (cardTypeId === DefaultCardTypes.KrDocTypeTypeID) {
      const sections = model.card.tryGetSections();
      if (sections) {
        const docTypeSection = sections.tryGet('KrDocType');
        const useForumBlock = model.blocks.get('UseForumBlock');
        if (docTypeSection && useForumBlock) {
          const useForumFirst = docTypeSection.fields.get(ForumHelper.UseForumField) as boolean;
          WfUiHelper.setControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForumFirst);

          const dispose = docTypeSection.fields.fieldChanged.addWithDispose(e => {
            if (e.fieldName === ForumHelper.UseForumField) {
              const useForum = e.fieldValue as boolean;
              e.storage.set(
                ForumHelper.UseDefaultDiscussionTabField,
                createTypedField(useForum, DotNetType.Boolean)
              );
              WfUiHelper.setControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForum);
            }
          });

          if (dispose) {
            this._disposers.push(dispose);
          }
        }
      }
    }
  }

  public finalized() {
    if (this._cardRowsListener) {
      this._cardRowsListener.stop();
      this._cardRowsListener = null;
    }
    for (let dispose of this._disposers) {
      if (dispose) {
        dispose();
      }
    }
    this._disposers.length = 0;
  }

  private attachHandlersToCardTypeRow(cardTypeRow: CardRow) {
    const dispose = cardTypeRow.fieldChanged.addWithDispose(e => {
      if (e.fieldName === ForumHelper.UseForumField && !(e.fieldValue as boolean)) {
        e.storage.set(
          ForumHelper.UseDefaultDiscussionTabField,
          createTypedField(e.fieldValue, DotNetType.Boolean)
        );
      }
    });

    if (dispose) {
      this._disposers.push(dispose);
    }
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/Forums/KrSettingsForumsSettingsUIExtension.cs
using System;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Client.UI;
using Tessa.Extensions.Default.Client.Workflow.Wf;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.Wf;
using Tessa.Forums;
using Tessa.Platform;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Default.Client.Forums
{
    /// <summary>
    /// Визуальное расширение для карточки настроек типов карточек и типа документов,
    /// связанных с системой форумов.
    /// </summary>
    public class KrSettingsForumsSettingsUIExtension : CardUIExtension
    {
        #region Private Methods

        private static void AttachHandlersToCardTypeRow(CardRow cardTypeRow)
        {
            cardTypeRow.FieldChanged += (s, e) =>
            {
                if (e.FieldName == ForumHelper.UseForumField
                    && !(bool)e.FieldValue)
                {
                    ((ICardFieldContainer)s).Fields[ForumHelper.UseDefaultDiscussionTabField] = BooleanBoxes.False;
                }
            };
        }

        #endregion

        #region Base Overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            ICardModel model = context.Model;
            if (model.InSpecialMode())
            {
                return;
            }

            Guid cardTypeID = model.CardType.ID;
            if (cardTypeID == DefaultCardTypes.KrSettingsTypeID)
            {
                StringDictionaryStorage<CardSection> sections = model.Card.TryGetSections();
                if (sections != null && sections.TryGetValue(WfHelper.CardTypeSettingsSection, out CardSection cardTypesSection) && model.Controls.TryGet(WfUIHelper.CardTypesControlName, out IControlViewModel cardTypesControl))
                {
                    // скрываем / отображаем контролы, связанные с использованием резолюции
                    var cardTypesGrid = (GridViewModel) cardTypesControl;
                    cardTypesGrid.RowInvoked += (s, e) =>
                    {
                        if ((e.Action == GridRowAction.Inserted || e.Action == GridRowAction.Opening) && e.RowModel.Blocks.TryGet(KrTypesUIHelper.UseForumBlock, out IBlockViewModel forumsBlock))
                        {
                            bool useForumFirst = e.Row.Get<bool>(ForumHelper.UseForumField);
                            WfUIHelper.SetControlVisibility(forumsBlock, ForumHelper.UseForumSuffix, useForumFirst);
                            EventHandler<CardFieldChangedEventArgs> fieldChangedHandler = (s2, e2) =>
                            {
                                if (e2.FieldName == ForumHelper.UseForumField)
                                {
                                    bool useForum = (bool) e2.FieldValue;

                                    ((ICardFieldContainer) s2).Fields[ForumHelper.UseDefaultDiscussionTabField] = useForum;
                                    WfUIHelper.SetControlVisibility(forumsBlock, ForumHelper.UseForumSuffix, useForum);
                                }
                            };

                            e.Row.FieldChanged += fieldChangedHandler;
                            forumsBlock.Form.Closed += (s2, e2) => e.Row.FieldChanged -= fieldChangedHandler;
                        }
                    };

                    // сбрасываем состояние полей, когда галка "использовать систему форумов" очищается
                    ListStorage<CardRow> cardTypeRows = cardTypesSection.Rows;
                    foreach (CardRow cardTypeRow in cardTypeRows)
                    {
                        AttachHandlersToCardTypeRow(cardTypeRow);
                    }

                    cardTypeRows.ItemChanged += (s, e) =>
                    {
                        // чтобы не было утечек памяти при удалении строки внутри обработчика нельзя использовать замыканий
                        if (e.Action == ListStorageAction.Insert)
                        {
                            AttachHandlersToCardTypeRow(e.Item);
                        }
                    };
                }
            }
            else if (cardTypeID == DefaultCardTypes.KrDocTypeTypeID)
            {
                StringDictionaryStorage<CardSection> sections = model.Card.TryGetSections();
                if (sections != null && sections.TryGetValue(WfHelper.DocTypeSettingsSection, out CardSection docTypeSection) && model.Blocks.TryGet(KrTypesUIHelper.UseForumBlock, out IBlockViewModel useForumBlock))
                {
                    // скрываем / отображаем контролы, связанные с использованием резолюции
                    bool useForumFirst = docTypeSection.RawFields.Get<bool>(ForumHelper.UseForumField);
                    WfUIHelper.SetControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForumFirst);

                    docTypeSection.FieldChanged += (s, e) =>
                    {
                        if (e.FieldName == ForumHelper.UseForumField)
                        {
                            // сбрасываем состояние полей, когда галка "использовать систему форумов" очищается
                            bool useForum = (bool) e.FieldValue;
                            
                            ((ICardFieldContainer) s).Fields[ForumHelper.UseDefaultDiscussionTabField] = useForum;
                            // скрываем / отображаем контролы, связанные с использованием резолюции
                            WfUIHelper.SetControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForum);
                        }
                    };
                }
            }
        }

        #endregion
    }
}